### PR TITLE
Add OpenCL backend solution adaptor

### PR DESCRIPTION
**Pull request by cgmillette**
- Add const_iterator object to KernelArguments to facilitate argument loading for openCL
- Add run-time check for 'argument logging' in KernelArguments when const_iterator is used (required)
- Add Solution Adapter object to load, launch and profile kernels in OpenCL environment.
- Add tests for new Solution Adapter
- Add cmake project option TENSILE_USE_OPENCL for using opencl
- Add tests to ensure HIP and openCL based solution adapters can be used at the same time
- Add OclAMDGPU object with query for minimal device properties, with tests comparing to HipAMDGPU

Tests:
RunGEMMKernelTests currently only uses the HIP device backend for setting up and running kernel tests.
Due to demand for an OpenCL backend, the test code needs to be a bit more flexible to be able to use OpenCL device backend.

- Add a layer of abstraction "DeviceBackend" to handle resource and device management, with an interface to populate contraction inputs, solve contraction problems and invoke kernels.

### CODE REVIEW

File: Tensile/Source/lib/source/KernelArguments.cpp [hunk 1]
```diff
@@ -120,4 +120,129 @@ namespace Tensile
     {
         return m_data.size();
     }
+
+    KernelArguments::const_iterator::const_iterator(KernelArguments const& args)
```
[the remark this answers is not in the document]
Comment (cgmillette) [reply Tensile/Source/lib/source/KernelArguments.cpp:124]: Done!

File: Tensile/Source/lib/include/Tensile/ocl/OclSolutionAdapter.hpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            SolutionAdapter(bool debug, std::string const& name);
+            SolutionAdapter(bool               debug,
+                            std::string const& name,
+                            cl::Context        context,
+                            cl::Device         device);
+            ~SolutionAdapter() final = default;
+
+            std::string name() const final;
+
+            void loadCodeObjectFile(std::string const& path);
```
[the remark this answers is not in the document]
Comment (cgmillette) [reply Tensile/Source/lib/include/Tensile/ocl/OclSolutionAdapter.hpp:53]: The PR was closed, so we can revisit this later.

**Review by pfultz2 — APPROVED** [review]
LGTM!

**Review by sdquiring — APPROVED** [review]

**Comment by sdquiring** [comment]
Should probably wait for extended tests to complete since they have been modified.

**Pull request merged**